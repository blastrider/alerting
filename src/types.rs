use std::fmt::{self, Display};
use std::str::FromStr;

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum AckFilter {
    Acked,
    Unacked,
    All,
}

impl AckFilter {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Acked => "ack",
            Self::Unacked => "unack",
            Self::All => "all",
        }
    }
}

impl Display for AckFilter {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for AckFilter {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_ascii_lowercase().as_str() {
            "ack" | "acked" => Ok(Self::Acked),
            "unack" | "unacked" => Ok(Self::Unacked),
            "all" => Ok(Self::All),
            other => Err(format!("unknown ack filter: {other}")),
        }
    }
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, Ord, PartialEq, PartialOrd, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Info,
    Warning,
    Average,
    High,
    Disaster,
}

impl Severity {
    pub fn from_zabbix(code: i64) -> Option<Self> {
        match code {
            1 => Some(Self::Info),
            2 => Some(Self::Warning),
            3 => Some(Self::Average),
            4 => Some(Self::High),
            5 => Some(Self::Disaster),
            _ => None,
        }
    }

    pub fn as_zabbix_code(self) -> i64 {
        match self {
            Self::Info => 1,
            Self::Warning => 2,
            Self::Average => 3,
            Self::High => 4,
            Self::Disaster => 5,
        }
    }
}

impl Display for Severity {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Severity::Info => "Info",
            Severity::Warning => "Warning",
            Severity::Average => "Average",
            Severity::High => "High",
            Severity::Disaster => "Disaster",
        })
    }
}

impl FromStr for Severity {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_ascii_lowercase().as_str() {
            "info" | "information" => Ok(Self::Info),
            "warn" | "warning" => Ok(Self::Warning),
            "average" => Ok(Self::Average),
            "high" => Ok(Self::High),
            "disaster" => Ok(Self::Disaster),
            other => Err(format!("unknown severity: {other}")),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{AckFilter, Severity};
    use std::str::FromStr;

    #[test]
    fn ack_filter_from_str_accepts_variants() {
        assert_eq!(AckFilter::from_str("ack"), Ok(AckFilter::Acked));
        assert_eq!(AckFilter::from_str("unacked"), Ok(AckFilter::Unacked));
        assert_eq!(AckFilter::from_str("ALL"), Ok(AckFilter::All));
        assert!(AckFilter::from_str("maybe").is_err());
    }

    #[test]
    fn severity_from_zabbix_parses_known_codes() {
        assert_eq!(Severity::from_zabbix(4), Some(Severity::High));
        assert_eq!(Severity::from_zabbix(1), Some(Severity::Info));
        assert!(Severity::from_zabbix(42).is_none());
    }
}
