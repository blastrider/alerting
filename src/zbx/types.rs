use serde::de::{self, Deserializer};
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Erreur Zabbix enveloppée par JSON-RPC.
#[derive(Debug, Deserialize)]
pub struct ZbxError {
    pub code: i64,
    pub message: String,
    #[serde(default)]
    pub data: Option<String>,
}

/// Enveloppe JSON-RPC générique.
#[derive(Debug, Deserialize)]
pub struct ZbxEnvelope<T> {
    #[allow(dead_code)]
    pub jsonrpc: String,
    pub result: Option<T>, // ne pas mettre #[serde(default)] ici
    pub error: Option<ZbxError>,
    #[allow(dead_code)]
    pub id: Value,
}

/// Problème Zabbix tel que renvoyé par `problem.get`.
#[derive(Debug, Deserialize)]
pub struct Problem {
    pub eventid: String,
    #[serde(deserialize_with = "de_i64_from_str")]
    pub clock: i64, // epoch sec en String -> i64
    #[serde(deserialize_with = "de_u8_from_str")]
    pub severity: u8, // "0".."5" -> u8
    pub name: String,
    #[serde(rename = "objectid")]
    #[allow(dead_code)]
    pub _objectid: Option<String>, // non utilisé dans l'affichage
    #[serde(default, deserialize_with = "de_bool_from_str_or_int")]
    pub acknowledged: bool, // "0"/"1" ou 0/1 -> bool
}

/// `event.get` avec hôtes.
#[derive(Debug, Deserialize)]
pub struct EventWithHosts {
    #[serde(default)]
    pub hosts: Vec<Host>,
}

#[derive(Debug, Deserialize)]
pub struct Host {
    #[serde(default)]
    pub host: Option<String>,
    #[serde(default)]
    pub name: Option<String>,
    // Zabbix peut renvoyer "0"/"1" (string) ou 0/1 (int) selon versions/widgets.
    #[serde(default, deserialize_with = "de_opt_u8_from_str_or_int")]
    pub status: Option<u8>, // 0 = enabled, 1 = disabled
}

/// Métadonnées d’hôte utilisées par le binaire (nom affichable + statut).
#[derive(Debug, Clone)]
pub struct HostMeta {
    pub display_name: String,
    pub status: Option<u8>,
}

/// Payload interne (privé au client) pour JSON-RPC.
#[derive(Serialize)]
pub(crate) struct RpcRequest<'a> {
    pub jsonrpc: &'static str,
    pub method: &'a str,
    pub params: Value,
    pub id: u32,
    pub auth: &'a str,
}

// ----------- helpers de désérialisation -----------

fn de_i64_from_str<'de, D>(deserializer: D) -> Result<i64, D::Error>
where
    D: Deserializer<'de>,
{
    let s = String::deserialize(deserializer)?;
    s.parse::<i64>()
        .map_err(|e| de::Error::custom(format!("clock invalide '{s}': {e}")))
}

fn de_u8_from_str<'de, D>(deserializer: D) -> Result<u8, D::Error>
where
    D: Deserializer<'de>,
{
    let s = String::deserialize(deserializer)?;
    s.parse::<u8>()
        .map_err(|e| de::Error::custom(format!("severity invalide '{s}': {e}")))
}

// --- helper: Option<u8> <- string/int/null ---
fn de_opt_u8_from_str_or_int<'de, D>(de: D) -> Result<Option<u8>, D::Error>
where
    D: Deserializer<'de>,
{
    #[derive(Deserialize)]
    #[serde(untagged)]
    enum U8OrStrOrNull {
        U8(u8),
        Str(String),
        Null,
    }
    match U8OrStrOrNull::deserialize(de)? {
        U8OrStrOrNull::U8(v) => Ok(Some(v)),
        U8OrStrOrNull::Str(s) => s.parse::<u8>().map(Some).map_err(de::Error::custom),
        U8OrStrOrNull::Null => Ok(None),
    }
}

fn de_bool_from_str_or_int<'de, D>(de: D) -> Result<bool, D::Error>
where
    D: Deserializer<'de>,
{
    #[derive(Deserialize)]
    #[serde(untagged)]
    enum Boolish {
        B(bool),
        U(u8),
        S(String),
        Null,
    }
    Ok(match Boolish::deserialize(de)? {
        Boolish::B(b) => b,
        Boolish::U(n) => n != 0,
        Boolish::S(s) => match s.as_str() {
            "1" | "true" | "True" | "TRUE" => true,
            _ => false,
        },
        Boolish::Null => false,
    })
}
