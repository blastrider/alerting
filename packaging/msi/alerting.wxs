<?xml version="1.0" encoding="UTF-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
  <Product
      Id="{5E5C86A5-2C30-4B37-8900-7C9D0E4D1A7A}"
      Name="Alerting"
      Language="1033"
      Version="1.0.0"
      Manufacturer="Alerting"
      UpgradeCode="{6D4F9B52-1AA1-4F9C-8AA2-B23D1562A21E}">
    <Package InstallerVersion="500" Compressed="yes" InstallScope="perUser" />

    <MajorUpgrade AllowDowngrades="no" />

    <MediaTemplate EmbedCab="yes" />

    <Directory Id="TARGETDIR" Name="SourceDir">
      <Directory Id="LocalAppDataFolder">
        <Directory Id="INSTALLFOLDER" Name="Alerting" />
      </Directory>
      <Directory Id="ProgramMenuFolder">
        <Directory Id="AlertingProgramMenu" Name="Alerting" />
      </Directory>
    </Directory>

    <Feature Id="Main" Level="1">
      <ComponentGroupRef Id="AlertingComponents" />
    </Feature>
  </Product>

  <Fragment>
    <DirectoryRef Id="INSTALLFOLDER">
      <Component Id="AlertingBinary" Guid="{8A9531A4-6D82-4F13-861D-0F8DC8D302FD}">
        <File Id="AlertingExe" Source="$(var.AlertingExecutable)" KeyPath="yes" />
      </Component>
    </DirectoryRef>

    <DirectoryRef Id="AlertingProgramMenu">
      <Component Id="StartMenuShortcut" Guid="{C0C072E7-8B46-4BDB-ADFB-99F36F7640F5}">
        <Shortcut Id="AlertingStartMenuShortcut" Name="Alerting" Target="[INSTALLFOLDER]alerting.exe" WorkingDirectory="INSTALLFOLDER" />
        <RegistryValue Root="HKCU" Key="Software\\Alerting" Name="StartMenuShortcut" Type="string" Value="1" KeyPath="yes" />
      </Component>
    </DirectoryRef>

    <ComponentGroup Id="AlertingComponents">
      <ComponentRef Id="AlertingBinary" />
      <ComponentRef Id="StartMenuShortcut" />
    </ComponentGroup>
  </Fragment>
</Wix>
