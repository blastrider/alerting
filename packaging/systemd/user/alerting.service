[Unit]
Description=Alerting Zabbix -> Desktop
After=network-online.target

[Service]
ExecStart=%h/.cargo/bin/alerting --config %h/.config/alerting/config.toml
Restart=on-failure
RestartSec=3
Environment=RUST_LOG=info
NoNewPrivileges=yes
PrivateTmp=yes
ProtectSystem=strict
ProtectHome=yes
ProtectControlGroups=yes
ProtectKernelTunables=yes
PrivateDevices=yes
LockPersonality=yes
MemoryDenyWriteExecute=yes
ReadWritePaths=%h/.config/alerting
SystemCallFilter=@system-service @pkey @chown @io-file

[Install]
WantedBy=default.target
