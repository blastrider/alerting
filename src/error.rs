use std::path::PathBuf;

use thiserror::Error;

#[derive(Debug, Error)]
pub enum Error {
    #[error(transparent)]
    Config(#[from] ConfigError),
    #[error(transparent)]
    Zabbix(#[from] ZbxError),
    #[error(transparent)]
    Notify(#[from] NotifyError),
    #[error("telemetry initialization failed: {0}")]
    Telemetry(String),
}

#[derive(Debug, Error)]
pub enum ConfigError {
    #[error("failed to read configuration file {path}")]
    FileRead {
        path: PathBuf,
        #[source]
        source: std::io::Error,
    },
    #[error("failed to parse configuration: {0}")]
    Parse(String),
    #[error("missing required configuration field: {field}")]
    MissingField { field: &'static str },
    #[error("invalid configuration for {field}: {message}")]
    InvalidField {
        field: &'static str,
        message: String,
    },
    #[error("configuration error: {0}")]
    Other(String),
}

#[derive(Debug, Error)]
pub enum ZbxError {
    #[error("failed to build HTTP client")]
    Client {
        #[source]
        source: reqwest::Error,
    },
    #[error("request failed: {source}")]
    Request {
        #[source]
        source: reqwest::Error,
    },
    #[error("unexpected HTTP status: {status}")]
    HttpStatus { status: reqwest::StatusCode },
    #[error("invalid JSON payload: {message}")]
    Json { message: String },
    #[error("invalid field {field}: {message}")]
    InvalidField {
        field: &'static str,
        message: String,
    },
    #[error("Zabbix API error {code}: {message}")]
    Api { code: i64, message: String },
    #[error("missing field in API response: {field}")]
    MissingField { field: &'static str },
    #[error("retry budget exhausted")]
    RetryExhausted {
        #[source]
        source: Box<ZbxError>,
    },
}

#[derive(Debug, Error)]
pub enum NotifyError {
    #[error("notification backend failed")]
    Backend,
    #[error("invalid notification payload: {0}")]
    InvalidPayload(String),
}

impl From<reqwest::Error> for ZbxError {
    fn from(source: reqwest::Error) -> Self {
        if source.is_status() {
            if let Some(status) = source.status() {
                return Self::HttpStatus { status };
            }
        }
        Self::Request { source }
    }
}

impl Error {
    pub fn is_retriable(&self) -> bool {
        matches!(
            self,
            Self::Zabbix(ZbxError::Request { .. })
                | Self::Zabbix(ZbxError::HttpStatus { .. })
                | Self::Zabbix(ZbxError::Json { .. })
        )
    }
}
