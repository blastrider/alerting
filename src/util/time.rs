use chrono::{Local, TimeZone};

/// Formatte un epoch (secondes) en heure locale lisible.
pub fn fmt_epoch_local(sec: i64) -> String {
    Local
        .timestamp_opt(sec, 0)
        .single()
        .map(|d| d.format("%Y-%m-%d %H:%M:%S").to_string())
        .unwrap_or_else(|| format!("(horodatage invalide: {sec})"))
}
