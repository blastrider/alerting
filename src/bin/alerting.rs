#[path = "alerting/app.rs"]
mod app;
#[path = "alerting/cli.rs"]
mod cli;
#[path = "alerting/notifier/mod.rs"]
mod notifier;
#[path = "alerting/rate_limit.rs"]
mod rate_limit;

use std::error::Error as StdError;

#[tokio::main]
async fn main() -> std::process::ExitCode {
    let cli = cli::Cli::parse_args();
    match app::run(cli).await {
        Ok(()) => std::process::ExitCode::SUCCESS,
        Err(err) => {
            report_error(&err);
            std::process::ExitCode::from(1)
        }
    }
}

fn report_error(err: &alerting::error::Error) {
    eprintln!("Error: {err}");
    let mut source: Option<&dyn StdError> = err.source();
    while let Some(cause) = source {
        eprintln!("  caused by: {cause}");
        source = cause.source();
    }
}
