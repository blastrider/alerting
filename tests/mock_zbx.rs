<<<<<<< HEAD
=======
#![allow(clippy::unwrap_used, clippy::expect_used)]

>>>>>>> feat/hardening-observability-ci
use std::time::Duration;

use alerting::error::{Error as AlertError, ZbxError};
use alerting::types::AckFilter;
use alerting::zbx_client::ZbxClient;
use secrecy::SecretString;
use serde_json::json;
use tokio::time::timeout;
use url::Url;
use wiremock::matchers::{body_string_contains, method};
use wiremock::{Mock, MockServer, ResponseTemplate};

fn client(base: &MockServer) -> ZbxClient {
    ZbxClient::new(
        Url::parse(&base.uri()).expect("valid mock url"),
        SecretString::from("token"),
        Duration::from_secs(2),
        Duration::from_secs(1),
        true,
    )
    .expect("client")
}

#[tokio::test]
async fn active_problems_returns_results() {
    let server = MockServer::start().await;

    Mock::given(method("POST"))
        .and(body_string_contains("problem.get"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "jsonrpc": "2.0",
            "result": [
                {
                    "eventid": "123",
                    "clock": "1700000000",
                    "lastchange": "1700000100",
                    "severity": "4",
                    "name": "Disk full",
                    "acknowledged": "0"
                }
            ],
            "id": 1
        })))
        .mount(&server)
        .await;

    Mock::given(method("POST"))
        .and(body_string_contains("event.get"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "jsonrpc": "2.0",
            "result": [
                {
                    "hosts": [
                        {
                            "host": "srv01",
                            "name": "Server 01",
                            "status": "0"
                        }
                    ]
                }
            ],
            "id": 1
        })))
        .mount(&server)
        .await;

    let client = client(&server);
    let problems = client
        .active_problems(10, AckFilter::All)
        .await
        .expect("problems");
    assert_eq!(problems.len(), 1);
    assert_eq!(problems[0].event_id, "123");
    assert!(!problems[0].acknowledged);

    let hosts = client
        .resolve_hosts(&["123".to_string()], 2)
        .await
        .expect("hosts");
    let meta = hosts[0].as_ref().expect("host meta");
    assert_eq!(meta.display_name, "Server 01");
}

#[tokio::test]
async fn problem_request_payload_snapshot() {
    let server = MockServer::start().await;

    Mock::given(method("POST"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "jsonrpc": "2.0",
            "result": [],
            "id": 1
        })))
        .mount(&server)
        .await;

    let client = client(&server);
    let _ = client.active_problems(5, AckFilter::Acked).await;

    let requests = server.received_requests().await.expect("requests");
    let body: serde_json::Value = serde_json::from_slice(&requests[0].body).expect("json body");
    insta::assert_json_snapshot!("problem_get_payload", body);
}

#[tokio::test]
async fn retries_exhaust_on_server_error() {
    let server = MockServer::start().await;

    Mock::given(method("POST"))
        .respond_with(ResponseTemplate::new(503))
        .mount(&server)
        .await;

    let client = client(&server);
    let err = client
        .active_problems(1, AckFilter::All)
        .await
        .expect_err("should fail");
    match err {
        AlertError::Zabbix(ZbxError::RetryExhausted { .. }) => {}
        other => panic!("unexpected error: {other}"),
    }
}

#[tokio::test]
async fn returns_api_error_details() {
    let server = MockServer::start().await;

    Mock::given(method("POST"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "jsonrpc": "2.0",
            "error": {
                "code": 42,
                "message": "Invalid token"
            },
            "id": 1
        })))
        .mount(&server)
        .await;

    let client = client(&server);
    let err = client
        .active_problems(1, AckFilter::All)
        .await
        .expect_err("should fail");
    match err {
        AlertError::Zabbix(ZbxError::Api { code, .. }) => assert_eq!(code, 42),
        _ => panic!("unexpected error"),
    }
}

#[tokio::test]
async fn timeouts_surface_as_errors() {
    let server = MockServer::start().await;

    Mock::given(method("POST"))
        .respond_with(
            ResponseTemplate::new(200)
                .set_body_json(json!({"jsonrpc":"2.0","result":[],"id":1}))
                .set_delay(Duration::from_millis(1500)),
        )
        .mount(&server)
        .await;

    let client = ZbxClient::new(
        Url::parse(&server.uri()).unwrap(),
        SecretString::from("token"),
        Duration::from_millis(500),
        Duration::from_millis(200),
        true,
    )
    .unwrap();

    let res = timeout(
        Duration::from_secs(5),
        client.active_problems(1, AckFilter::All),
    )
    .await;
    let err = res.expect("timeout future").expect_err("should fail");
    assert!(matches!(err, AlertError::Zabbix(ZbxError::Request { .. })));
}
