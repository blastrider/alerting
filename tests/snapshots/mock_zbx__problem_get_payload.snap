---
source: tests/mock_zbx.rs
assertion_line: 102
expression: body
---
{
  "auth": "token",
  "id": 1,
  "jsonrpc": "2.0",
  "method": "problem.get",
  "params": {
    "acknowledged": true,
    "limit": 5,
    "output": [
      "eventid",
      "name",
      "severity",
      "clock",
      "lastchange",
      "acknowledged"
    ],
    "recent": false,
    "sortfield": [
      "eventid"
    ],
    "sortorder": "DESC"
  }
}
